c10/core/DispatchKey.h

```diff
@@ -133,6 +133,7 @@ enum class DispatchKey : uint8_t {
   BackendSelect,
 
   Python,
+  PythonMode,
```

Comment:
I have a more general question - what do you think is the correct interaction between this new feature and the previous usages of `__torch_dispatch__`? Are they meant to be used together? E.g.
```
x = SubclassA(torch.randn([]))
y = SubclassA(torch.randn([]))
        with enable_python_mode(SubclassB):
            z = x + y
            # I think z is a subclassB?
            
```

Part of the reason that I'm wondering is because it looks like `Python` and `PythonMode` share a lot of the same logic. Is there something stopping us from merging them? I guess we would have to make `Python` sometimes function as a modal key, for one. And we'd want to make sure that the `TorchDispatchTypeObject` gets priority when choosing what subclass to dispatch to (which I think it already does)

Reply:
Yes, they are meant to be used together. The use case of that is if we have something like LoggingTensor (or FiniteTensor), then we may want to log ALL tensor operations, even the ones that happen outside of LoggingTensor.

> Part of the reason that I'm wondering is because it looks like Python and PythonMode share a lot of the same logic. Is there something stopping us from merging them?

Now that you mention it, I think we could just combine the two keys. I didn't have a good reason for separating them; I remember Ed proposed a trick for combining Batched and VmapMode and I didn't want to go through that exercise but we don't need to do that here

Reply:
>  I guess we would have to make Python sometimes function as a modal key, for one.

Yup

> And we'd want to make sure that the TorchDispatchTypeObject gets priority when choosing what subclass to dispatch to (which I think it already does)

Yup, that already happens

Reply:
Resolution: Python and PythonMode were collapsed into a single key